### PR TITLE
Add option to use closure for PageCache $variations

### LINKED ISSUE #19740: PageCache variations as callable
In order to access the `Yii::$app->user` inside a PageCache variation situation, the variations needs to be a callable. Imagine this situation:

```php
public function behaviors()
{
    return [
        'authenticator' => [
            'class' => CompositeAuth::class,
        ],
        'pageCache' => [
            'class' => PageCache::class,
            'variations' => [
                Yii::$app->user->identitly->prop_user_id,
            ]
        ]
    ];
}
```

This will throw an exception, since `Yii::$app->user->identitly->prop_user_id` is evaulated at runtime while configure the behaviors. But since the Authenticator is running before the pageCache behavior, the user id would be available. In order to fix that `variations` should allow a closure:

```php
public function behaviors()
{
    return [
        'authenticator' => [
            'class' => CompositeAuth::class,
        ],
        'pageCache' => [
            'class' => PageCache::class,
            'variations' => function() {
                return Yii::$app->user->identitly->prop_user_id,
            }
        ]
    ];
}
```

If https://github.com/yiisoft/yii2/blob/master/framework/filters/PageCache.php#L315 `calculateCacheKey` would not be protected, i could at least override the PageCache with my own implementation, but its protected. So i assume the above integration of checking for a closure/callable is the only solution.

If there is another solution to allow full page cache based on the current user, let me know. Afaik, there is no solution, otherwise this would be resolved https://github.com/yiisoft/yii2/issues/19120

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

 I would send a PR if its accepted.

### PR DESCRIPTION
**Pull request by nadar**
this closes https://github.com/yiisoft/yii2/issues/19740

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #19740

### CODE REVIEW

File: framework/filters/PageCache.php [hunk 1]
```diff
@@ -108,6 +109,20 @@ class PageCache extends ActionFilter implements DynamicContentAwareInterface
...
+     * ```php
+     * 'variations' => function() {
+     *     return [
+     *         Yii::$app->language,
+     *         Yii::$app->user->id
+     *     ];
+     * }
+     * ```
+     *
+     * The callable should return an array otherwise it will be type casted to an array.
```
Comment (bizley) [inline framework/filters/PageCache.php:125]:
```suggestion
     * The callable should return an array.
```

**Comment by nadar** [comment]
Thanks @bizley

**Comment by nadar** [comment]
+ tests are failing, but seems not cause of my code
+ is there any standard naming? Is it a callable, closure or anonymous function
+ Shall we add Closure to the use section or \Closure as inline?

**Comment by bizley** [comment]
Yes, we see segmentation faults for some time. I think you should add `if ($this->variations instanceof Closure || (is_array($this->variations) && is_callable($this->variations))) {`, which would be in line with the rest of the framework. We import Closure usually I think (not inlined).

**Comment by nadar** [comment]
`(is_array($this->variations) && is_callable($this->variations))` will never evaluted true i assume...? afaik `callable` should not be used because it can trigger internal methods?

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Comment by rob006** [comment]
Treating array as callable is IMO a big mistake, and possibly security issue. Array is simply ambiguous in this context - we should never treat array as callable in context, when raw array is also a valid value. In this case `['Foo', 'bar']` will have different meaning depending on whether `Foo::bar()` exists. And since `$variations` often depends on some data parameters from request, this could be used to run some unexpected code on server using malformed request. For example previously perfectly safe `[$_GET['param1'] ?? null, $_GET['param2'] ?? null]` now could be used to run any static method in framework.

**Comment by nadar** [comment]
So lets remove the array option and just keep the `instanceof Closure` check :+1:

**Comment by nadar** [comment]
@bizley is there any ETA for the next release? It's almost 6 months since last release, we are stuck on dev-master :cry:

**Comment by bizley** [comment]
It's the Samdark's decision.

**Pull request merged**